crates/amaru-sim/src/echo/run.rs

```diff
@@ -0,0 +1,295 @@
+    send_output.upstream.connect(from_echo);
+
+    let policy = define_gasket_policy();
+
+    let read_input = gasket::runtime::spawn_stage(read_input, policy.clone());
+    let echo = gasket::runtime::spawn_stage(echo, policy.clone());
+    let send_output = gasket::runtime::spawn_stage(send_output, policy.clone());
+
+    vec![read_input, echo, send_output]
+}
+
+pub async fn run_pipeline(pipeline: gasket::daemon::Daemon, exit: CancellationToken) {
+    loop {
+        tokio::select! {
+            _ = tokio::time::sleep(Duration::from_secs(1000)) => {
+                if pipeline.should_stop() {
+                    break;
+                }
+            }
+            _ = exit.cancelled() => {
+                trace!("exit requested");
+                break;
+            }
+        }
+    }
+
+    trace!("shutting down pipeline");
+
+    pipeline.teardown();
+}
+
+fn define_gasket_policy() -> gasket::runtime::Policy {
+    let retries = gasket::retries::Policy {
+        max_retries: 20,
+        backoff_unit: std::time::Duration::from_secs(1),
+        backoff_factor: 2,
+        max_backoff: std::time::Duration::from_secs(60),
+        dismissible: false,
+    };
+
+    gasket::runtime::Policy {
+        //be generous with tick timeout to avoid timeout during block awaits
+        tick_timeout: std::time::Duration::from_secs(600).into(),
+        bootstrap_retry: retries.clone(),
+        work_retry: retries.clone(),
+        teardown_retry: retries.clone(),
+    }
+}
+
+type EchoesOut = gasket::messaging::OutputPort<Envelope>;
+type EchoesIn = gasket::messaging::InputPort<Envelope>;
+
+struct Action {}
+
+struct ReadWorker {}
+
+#[derive(Stage)]
+#[stage(name = "read", unit = "Action", worker = "ReadWorker")]
+struct ReadInput {
+    downstream: EchoesOut,
+}
+
+impl ReadInput {
+    fn new() -> Self {
+        Self {
+            downstream: EchoesOut::default(),
+        }
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl gasket::framework::Worker<ReadInput> for ReadWorker {
+    async fn bootstrap(_stage: &ReadInput) -> Result<Self, WorkerError> {
+        let worker = Self {};
+
+        Ok(worker)
+    }
+
+    async fn schedule(
+        &mut self,
+        _stage: &mut ReadInput,
+    ) -> Result<WorkSchedule<Action>, WorkerError> {
+        Ok(WorkSchedule::Unit(Action {}))
+    }
+
+    async fn execute(&mut self, _unit: &Action, stage: &mut ReadInput) -> Result<(), WorkerError> {
+        let mut input = String::new();
+        io::stdin()
+            .read_line(&mut input)
+            .map_err(|_| WorkerError::Recv)?;
+
+        let msg: Envelope = serde_json::from_str(&input).map_err(|_| WorkerError::Recv)?;
+
+        trace!("message '{:?}'", msg);
+
+        stage
+            .downstream
+            .send(msg.into())
+            .await
+            .map_err(|_| WorkerError::Panic)?;
+        Ok(())
+    }
+}
+
+#[derive(Stage)]
+#[stage(name = "do_echo", unit = "Action", worker = "EchoWorker")]
+struct DoEcho {
```

Comment:
There are three, largely orthogonal, things going on in this module:

  1. Gasket related things (down/upstream);
  2. Node related things (node_id and count);
  3. The echo "business logic".

It could worth to try to isolate them from each other.

Reply:
Agreed. The stage derivation is somewhat annoying, I will see if I can move it out of the way.